ykrt/src/compile/jitc_yk/opt/analyse.rs

```diff
@@ -0,0 +1,82 @@
+//! Analyse a trace and gradually refine what values we know a previous instruction can produce.
+
+use super::{
+    super::jit_ir::{GuardInst, Inst, InstIdx, Module, Operand, Predicate},
+    Value,
+};
+
+/// Ongoing analysis of a trace: what value can a given instruction in the past produce?
+///
+/// Note that the analysis is forward-looking: just because an instruction's `Value` is (say) a
+/// `Const` now does not mean it would be valid to assume that at earlier points it is safe to
+/// assume it was also a `Const`.
+pub(super) struct Analyse {
+    /// For each instruction, what have we learnt about its [Value] so far?
+    values: Vec<Value>,
+}
+
+impl Analyse {
+    pub(super) fn new(m: &Module) -> Analyse {
+        Analyse {
+            values: vec![Value::Unknown; m.insts_len()],
+        }
+    }
+
+    /// Map `op` based on our analysis so far. In some cases this will return `op` unchanged, but
+    /// in others it may be able to turn what looks like a variable reference into a constant.
+    pub(super) fn op_map(&mut self, m: &Module, op: Operand) -> Operand {
+        match op {
+            Operand::Var(iidx) => match self.values[usize::from(iidx)] {
+                Value::Unknown => {
+                    // Since we last saw an `ICmp` instruction, we may have gathered new knowledge
+                    // that allows us to turn it into a constant.
+                    if let (iidx, Inst::ICmp(inst)) = m.inst_decopy(iidx) {
+                        let lhs = self.op_map(m, inst.lhs(m));
+                        let pred = inst.predicate();
+                        let rhs = self.op_map(m, inst.rhs(m));
+                        if let (&Operand::Const(lhs_cidx), &Operand::Const(rhs_cidx)) = (&lhs, &rhs)
+                        {
+                            if pred == Predicate::Equal && m.const_(lhs_cidx) == m.const_(rhs_cidx)
+                            {
+                                self.values[usize::from(iidx)] = Value::Const(m.true_constidx());
+                                return Operand::Const(m.true_constidx());
+                            }
+                        }
+                    }
+                    op
+                }
+                Value::Const(cidx) => Operand::Const(cidx),
```

Comment:
Potential future work out of scope of this PR.

Instead of tracking a single constant value, you could plug other domain(s) so that you could track richer information and use the analysis more widely for other optimisations besides just constant propagation.

For example an interval domain to express ranges, or a relational domain to express that `%2 = %1 + 1`, or `%2 < %1`.

Reply:
Yes, that's what RPython does, but one thing at a time.